### PR TITLE
deadlock: speed up detecting leader change of the deadlock detector

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

## What have you changed? (mandatory)
To speed up detecting role change of the deadlock detector, if the node has the region containing key `b''`, the node will get leader change info from raftstore.  It the only way to change from leader to follower, and vice versa.

The timer to monitor membership change is removed. Now if the connection to the leader is disconnected when trying to send a detect request, try to get the latest leader info from pd immediately.

I found that it is difficult to modify based on the original implementation, so I moved almost fields of `Inner` to `Detector`.

## What are the type of the changes? (mandatory)
- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
Manual tests.

## Does this PR affect documentation (docs) or release note? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
Yes. I will file a PR to update it.

### CODE REVIEW

**Comment by youjiali1995** [comment]
@coocood @zhangjinpeng1987 @AndreMouche @MyonKeminta PTAL

File: src/storage/lock_manager/deadlock.rs [hunk 1]
```diff
@@ -6,26 +6,37 @@ use super::util::extract_physical_timestamp;
...
+/// The leader of the region containing the LEADER_KEY is the leader of deadlock detector.
+const LEADER_KEY: &[u8] = b"";
+
+/// Returns true if the region containing the LEADER_KEY.
+fn is_leader_region(region: &'_ Region) -> bool {
+    region.get_start_key() <= LEADER_KEY && LEADER_KEY < region.get_end_key()
+}
+
 // 2 mins
 const TXN_DETECT_INFO_TTL: u64 = 120000;
```
Comment (coocood) [inline src/storage/lock_manager/deadlock.rs]: This value can be reduced?

Comment (youjiali1995) [reply src/storage/lock_manager/deadlock.rs]: See https://github.com/tikv/tikv/pull/5028/files#diff-fa2bb0fa5ba566c8685b2a159a41ada6R291
If the performance is good, the timer will be removed. If not, I will file a PR to change it.

**Comment by youjiali1995** [comment]
image

I found a race condition:
1. The node is the leader and changes to the follower.
2. It receives the role change event from raftstore first, and changes to the follower.
3. There is a detect request to send, the node will refresh leader info but the Pd thought this node is the leader due to stale info. The node becomes the leader again.
4. The only way to change a node from leader to follower is notified by raftstore. It's possible the step 2 in the image happens after step 3.

I will fix it.

File: src/storage/lock_manager/deadlock.rs [hunk 2]
```diff
@@ -391,9 +394,9 @@ impl<S: StoreAddrResolver + 'static> Detector<S> {
     fn change_role(&mut self, role: StateRole) {
         if self.inner.borrow().role != role {
             if role == StateRole::Leader {
-                info!("became the leader of deadlock detector!");
+                info!("became the leader of deadlock detector!"; "self_id" => self.store_id);
             } else {
-                info!("changed from leader to follower");
+                info!("changed from leader to follower"; "self_id" => self.store_id);
```
Comment (coocood) [inline src/storage/lock_manager/deadlock.rs]: add `of deadlock detector`?

**Comment by coocood** [comment]
LGTM

**Comment by AndreMouche** [comment]
Any update?

**Comment by youjiali1995** [comment]
@AndreMouche @MyonKeminta @sticnarf PTAL

**Comment by youjiali1995** [comment]
I tested it with https://github.com/pingcap/schrodinger-test/pull/386. And I set `wait-for-lock-timeout` to `100ms` which affects the frequency of retry.

The metric is how long it takes to detect the deadlock when the leader of deadlock detector is in chaos. As the metric shows, the max duration is about `wait-for-lock-timeout`, because if the detect request is sent to the wrong leader, the next detect request won't be sent until waiter timeout. But when it retries, it can send to the right leader.
image

**Comment by youjiali1995** [comment]
@AndreMouche @MyonKeminta PTAL

File: src/storage/lock_manager/deadlock.rs [hunk 3]
```diff
@@ -5,36 +5,45 @@ use super::config::Config;
...
+use tikv_util::worker::{FutureRunnable, FutureScheduler, FutureWorker, Stopped};
 use tokio_core::reactor::Handle;
-use tokio_timer::Interval;
+
+/// The leader of the region containing the LEADER_KEY is the leader of deadlock detector.
+const LEADER_KEY: &[u8] = b"";
+
+/// Returns true if the region containing the LEADER_KEY.
+fn is_leader_region(region: &'_ Region) -> bool {
+    region.get_start_key() <= LEADER_KEY && LEADER_KEY < region.get_end_key()
```
Comment (MyonKeminta) [inline src/storage/lock_manager/deadlock.rs]: When talking about end_key, we use an empty key to represent the maximum key (∞). So when you want to check if a key is less than a region's end_key, you checks `region.get_end_key().is_empty() || key < region.get_end_key()`.
```suggestion
    region.get_start_key() <= LEADER_KEY && (region.get_end_key().is_empty() || LEADER_KEY < region.get_end_key())
```

File: src/storage/lock_manager/deadlock.rs [hunk 4]
```diff
@@ -371,162 +506,101 @@ impl Inner {
...
             for _ in 0..2 {
-                if inner.leader_client.is_none() && inner.leader_info.is_some() {
-                    inner.reconnect_leader(handle);
+                // TODO: If the leader hasn't been elected, it requests Pd for
+                // each detect request. Maybe need flow control here.
+                //
+                // Refresh leader info when the connection to the leader is disconnected.
+                if self.leader_client.is_none() && !self.refresh_leader_info() {
+                    // Return if the leader doesn't exist.
+                    return;
```
Comment (MyonKeminta) [inline src/storage/lock_manager/deadlock.rs]: The detect request will be directly ignored here? May it cause any problem? Should here be a warn?

File: src/storage/lock_manager/deadlock.rs [hunk 5]
```diff
@@ -278,79 +299,193 @@ impl Scheduler {
...
+impl RoleObserver for Scheduler {
+    fn on_role_change(&self, ctx: &mut ObserverContext<'_>, role: StateRole) {
+        if is_leader_region(ctx.region()) {
+            self.change_role(role)
+        }
+    }
+}
+
+/// Creates the `Scheduler` and registers it to the CoprocessorHost.
+pub fn register_role_change_observer(host: &mut CoprocessorHost, worker: &FutureWorker<Task>) {
```
[the remark this answers is not in the document]
Comment (youjiali1995) [reply src/storage/lock_manager/deadlock.rs]: I renamed it to `register_detector_role_change_observer`.

File: src/storage/lock_manager/deadlock.rs [hunk 6]
```diff
@@ -278,79 +299,193 @@ impl Scheduler {
...
+    let scheduler = Scheduler::new(worker.scheduler());
+    host.registry.register_role_observer(1, Box::new(scheduler));
 }
 
-#[derive(Clone)]
 struct Inner {
+    /// The role of the deadlock detector. Default is `StateRole::Follower`.
+    role: StateRole,
```
[the remark this answers is not in the document]
Comment (youjiali1995) [reply src/storage/lock_manager/deadlock.rs]: Done.

File: src/storage/lock_manager/deadlock.rs (hunk 4 above)
Comment (youjiali1995) [reply src/storage/lock_manager/deadlock.rs]: It's a bug. I changed the `return` to `break` to log it.

File: src/storage/lock_manager/deadlock.rs [hunk 7]
```diff
@@ -278,79 +325,196 @@ impl Scheduler {
...
+
+impl Coprocessor for Scheduler {}
+
+/// Implements observer traits for `Scheduler`.
+/// If the role of the node in the leader region changes, notifys the deadlock detector.
+///
+/// If the leader region is merged or splited in the node, the role of the node won't change.
+/// If the leader region is removed and the node is the leader, it will change to follower first.
+/// So there is no need to observe region change events.
+impl RoleObserver for Scheduler {
```
Comment (AndreMouche) [inline src/storage/lock_manager/deadlock.rs]: Could we move this part into `rafstore/coprocessor`? Seems all observers are in it

Comment (youjiali1995) [reply src/storage/lock_manager/deadlock.rs]: I don't like it. Maybe I will do it when separates storage as a component.

Comment (MyonKeminta) [reply src/storage/lock_manager/deadlock.rs]: I don't think it's a good idea too. Raftstore provides the Observer traits, and it should be implemented where it's used.

**Pull request merged**